### DISCUSSION TITLE
Best place to start implementing OVERLAPS?

### DISCUSSION THREAD

**Discussion by d2a-raudenaerde** [JSQLParser/JSqlParser#1600] [Q&A] [answered]
I'm trying to implement OVERLAPS, which looks like this:

```
(TIME '2:55:00', INTERVAL '1' HOUR)
OVERLAPS
(TIME '3:30:00', INTERVAL '2' HOUR)

```
I think the OverlapsCondition would look like this:
` "(" SimpleExpression() "," SimpleExpression() ")" <K_OVERLAPS> "(" SimpleExpression() "," SimpleExpression() ")"`
where the expressions are stored as peroid1start, peroid1end, peroid2start, peroid2end.

However, I find it a bit difficult to find the right place to implement this in, because it is a bit of a different type than regular operators, like <, > etc.

I currently think the best place is in:

`Condition():` 

Where it would be a 3th branch next to the 
```
    (
        LOOKAHEAD(RegularCondition()) result=RegularCondition()
        | result=SQLCondition()
    )
```

`LOOKAHEAD(OverlapsCondition()) result = OverlapsCondition()`

However, I'm not really sure about this approach. 
WDYT?

**Comment by d2a-raudenaerde** [comment]
Ok while that works for my simple use-case, I looked it up in the sql2003 standard:

There it is:
`<row value predicand 1> OVERLAPS <row value predicand 2>`,

Where 
```
<row value predicand> ::=
		<row value special case>
	|	<row value constructor predicand>
```

And 
```
<row value constructor predicand> ::=
		<common value expression>
	|	<boolean predicand>
	|	<explicit row value constructor>
```
And
```
<explicit row value constructor> ::=
		<left paren> <row value constructor element> <comma> <row value constructor element list> <right paren>
	|	ROW <left paren> <row value constructor element list> <right paren>
	|	<row subquery>
```

Which makes sense. 

So I tried to refactor to:

`RowConstructor() <K_OVERLAPS> RowConstructor()`

But the `RowConstructor()` does not accept stuff like: 

```
        String sql = "(t1.col1, t1.col2)";
        CCJSqlParser parser = new CCJSqlParser(new StringProvider(sql)).withAllowComplexParsing(false);
        Expression e =  parser.RowConstructor();
        System.out.println(e.toString());
```

Any hints?

**Reply by d2a-raudenaerde** [reply]
`RowConstructor` probably has to be split into `RowValueConstructor` and `RowTypeConstructor`, where the latter is used in `CastExpression()`

**Comment by manticore-projects** [comment, marked as answer]
As far as I do understand JSQLParser, our best match for `<row value constructor predicand>` was `Expression`.
I think you are looking for something like:
```
"(" ExpressionList() ")" <OVERLAPS> "(" ExpressionList() ")"
```

Please refer to JSQLParser/src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt:3580
Alternatively JSQLParser/src/main/jjtree/net/sf/jsqlparser/parser/JSqlParserCC.jjt:3595

**Reply by d2a-raudenaerde** [reply]
Ah, great pointers.

**Reply by manticore-projects** [reply]
Rob,

please let me apologize, I should have pointed that out earlier. I know how frustrating it is to work on such stuff and be forced to move it into a different direction after. I will try to respond more timely, sorry again.

**Reply by d2a-raudenaerde** [reply]
Thanks and no worries, I cannot expect ppl to react immediately :)

**Reply by d2a-raudenaerde** [reply]
I have created a new pull-request incorporating all feedback, I think it is much nicer as well like this!
